cluster-autoscaler/simulator/cluster_snapshot_test.go

```diff
@@ -0,0 +1,435 @@
+				err := snapshot.AddNode(node)
+				assert.NoError(t, err)
+			},
+			modifiedState: snapshotState{
+				nodes: []*apiv1.Node{node},
+			},
+		},
+		{
+			name: "add node with pods",
+			op: func(snapshot ClusterSnapshot) {
+				err := snapshot.AddNodeWithPods(node, []*apiv1.Pod{pod})
+				assert.NoError(t, err)
+			},
+			modifiedState: snapshotState{
+				nodes: []*apiv1.Node{node},
+				pods:  []*apiv1.Pod{pod},
+			},
+		},
+		{
+			name: "remove node",
+			state: snapshotState{
+				nodes: []*apiv1.Node{node},
+			},
+			op: func(snapshot ClusterSnapshot) {
+				err := snapshot.RemoveNode(node.Name)
+				assert.NoError(t, err)
+			},
+		},
+		{
+			name: "remove node, then add it back",
+			state: snapshotState{
+				nodes: []*apiv1.Node{node},
+			},
+			op: func(snapshot ClusterSnapshot) {
+				err := snapshot.RemoveNode(node.Name)
+				assert.NoError(t, err)
+
+				err = snapshot.AddNode(node)
+				assert.NoError(t, err)
+			},
+			modifiedState: snapshotState{
+				nodes: []*apiv1.Node{node},
+			},
+		},
+		{
+			name: "add pod, then remove node",
+			state: snapshotState{
+				nodes: []*apiv1.Node{node},
+			},
+			op: func(snapshot ClusterSnapshot) {
+				err := snapshot.AddPod(pod, node.Name)
+				assert.NoError(t, err)
+				err = snapshot.RemoveNode(node.Name)
+				assert.NoError(t, err)
+			},
+		},
+	}
+
+	for name, snapshotFactory := range snapshots {
+		for _, tc := range testCases {
+			t.Run(fmt.Sprintf("%s: %s base", name, tc.name), func(t *testing.T) {
+				snapshot := startSnapshot(t, snapshotFactory, tc.state)
+				tc.op(snapshot)
+
+				// Modifications should be applied.
+				compareStates(t, tc.modifiedState, getSnapshotState(t, snapshot))
+			})
+			t.Run(fmt.Sprintf("%s: %s fork", name, tc.name), func(t *testing.T) {
+				snapshot := startSnapshot(t, snapshotFactory, tc.state)
+
+				err := snapshot.Fork()
+				assert.NoError(t, err)
+
+				tc.op(snapshot)
+
+				// Modifications should be applied.
+				compareStates(t, tc.modifiedState, getSnapshotState(t, snapshot))
+			})
+			t.Run(fmt.Sprintf("%s: %s fork & revert", name, tc.name), func(t *testing.T) {
+				snapshot := startSnapshot(t, snapshotFactory, tc.state)
+
+				err := snapshot.Fork()
+				assert.NoError(t, err)
+
+				tc.op(snapshot)
+
+				err = snapshot.Revert()
+				assert.NoError(t, err)
+
+				// Modifications should no longer be applied.
+				compareStates(t, tc.state, getSnapshotState(t, snapshot))
+			})
+			t.Run(fmt.Sprintf("%s: %s fork & commit", name, tc.name), func(t *testing.T) {
+				snapshot := startSnapshot(t, snapshotFactory, tc.state)
+
+				err := snapshot.Fork()
+				assert.NoError(t, err)
+
+				tc.op(snapshot)
+
+				err = snapshot.Commit()
+				assert.NoError(t, err)
+
+				// Modifications should be applied.
+				compareStates(t, tc.modifiedState, getSnapshotState(t, snapshot))
+			})
+		}
+	}
+}
+
+func TestClear(t *testing.T) {
+	// Run with -count=1 to avoid caching.
+	localRand := rand.New(rand.NewSource(time.Now().Unix()))
```

Comment:
I have mixed feelings about randomized ut, but in this case I can't name a good argument against it (you log all randomly generated numbers, so failures are easily reproducible). So I guess it's fine.